src/passes/OptimizeInstructions.cpp

```diff
@@ -1391,6 +1397,83 @@ struct OptimizeInstructions
     return binary;
   }
 
+  Expression* optimizeBulkMemory(Expression* curr) {
+    if (auto* memCopy = curr->dynCast<MemoryCopy>()) {
+      FeatureSet features = getModule()->features;
+
+      auto destHasSideEffects =
+        EffectAnalyzer(getPassOptions(), features, memCopy->dest)
+          .hasSideEffects();
+
+      // memory.copy(x, x, sz)  ==>  nop
+      if (!destHasSideEffects &&
+          ExpressionAnalyzer::equal(memCopy->dest, memCopy->source)) {
+        return ExpressionManipulator::nop(curr);
+      }
+      // memory.copy(dst, src, C)  ==>  store(dst, load(src))
+      if (auto* csize = memCopy->size->dynCast<Const>()) {
+        auto bytes = csize->value.geti32();
+
+        switch (bytes) {
+          case 0: {
+            if (!destHasSideEffects &&
+                !EffectAnalyzer(getPassOptions(), features, memCopy->source)
+                   .hasSideEffects()) {
+              return ExpressionManipulator::nop(curr);
+            }
+            break;
+          }
+          case 1:
+          case 2:
+          case 4: {
+            Builder builder(*getModule());
+            return builder.makeStore(
+              bytes, // bytes
+              0,     // offset
+              bytes, // align
```

Comment:
It would be more conservative to use an alignment of 1 for these loads and stores since we don't know whether the original memory.copy was aligned.